### ISSUE TITLE
Co-host has no privileges

### ISSUE THREAD

**Issue by SnacksonMA** [ncstate-delta/moodle-mod_zoom#254]
Hi. There is a problem with co-hosting. If the user (Teacher role) is a Zoom pro and creates a zoom session for himself, it works fine. But If the user creates a zoom session with a Co-host (also Zoom pro), the Co-host don't get host privileges. We have Moodle 3.9.6 / Plugin Zoom v3.6. :)

**Comment by jrchamp** [comment]
Zoom overloads a lot of these words, so I'm going to need to over-explain (sorry!).

The user that creates the activity is the "outside of the meeting" Host. Also, outside of the meeting, you can add another user as an Alternative Host. Inside of the meeting, the first person between these two to join becomes the "inside the meeting" Host. When the "outside of the meeting" Host joins, depending on how Zoom is behaving that day, they either become the "inside the meeting" Host automatically or have the option to "Claim Host". Sometimes, this results in the "Alternative Host" user who was the "inside the meeting" Host to become an "inside the meeting" Co-Host or a basic Participant. There can only be 1 "inside the meeting" Host at a time (a Zoom restriction). To automatically get "inside the meeting" Host or Co-Host access, you must be logged into the Zoom client using the account that has "outside the meeting" Host or Alternative Host.

Given all that, as Zoom is a cloud service, if this functionality works for *anyone* then it should work for *everyone* who has configured the Zoom Meeting and is logged in, as described. It's possible that you've found a bug or that one of the described pre-conditions was not met.

One way to check that Alternative Host is set correctly is to go to the Zoom web interface directly. That would help diagnose if there is a mismatch between what is showing in Moodle and what Zoom's portal believes is correct.

Again, sorry for the wall of text - I overshare a bit.

**Comment by SnacksonMA** [comment]
Hi Jonathan. Thanks for your explanation.

We are using both methods to connect with zoom. LTI and your Plugin. The LTI is working fine with an alternative host. 

I tried to access using the web inferface, and doesn't work. The teacher (alternative host) acts as student.

-- 

I tried:

User Zoom 1
User Zoom 2

When you take a look at the URL that generates the Zoom, there is differences:

A) Room created by another Zoom user 1 -- Assigned to User Zoom 2 (both  ZOOM PRO / Teacher)
Ask for CODE -- Logged in as a student
https://alku-videoclases-es.zoom.us/j/91535005282?uname=Profe%20Zigurat#success

B) Room created by another Zoom user 2 directly (ZOOM PRO / Teacher)
Works fine!
https://alku-videoclases-es.zoom.us/s/98974917550#success

It's like the user is not been transferred to Zoom, unless you are the same user than creates the room in Moodle.

:)

**Comment by jrchamp** [comment]
Hi @SnacksonMA 

To check the meeting settings, here is the Meeting information page: https://alku-videoclases-es.zoom.us/meeting/91535005282

You should see:

> Alternative Hosts: User Zoom 2 (user-zoom2@example.com)

**Comment by SnacksonMA** [comment]
Hi Jonathan

Yes, it is. 

![image](https://user-images.githubusercontent.com/83217628/116511198-5d76f480-a8c6-11eb-9006-03c47d3ba015.png)

:)

**Comment by jrchamp** [comment]
If the user is present in the Alternative Hosts list, then that should mean it is set up correctly as far as the plugin is concerned. Please make sure that the profezig@snackson.com user is logged in to the Zoom client.

**Comment by jrchamp** [comment]
No, the plugin does not send the credentials through the app. It is not possible, because it is a security risk. Zoom only trusts Zoom, Zoom doesn't trust us or anyone else. That is a good thing, because otherwise we could pretend to be anyone and take over control of their account.

**Comment by SnacksonMA** [comment]
Hi. Sorry but I believe that premise is wrong. 

Teacher 1 (PRO ZOOM) creates a Zoom session. It works fine. Zoom sends the user info (not the credentials) of the Zoom user to the app (same mail in both Moodle and Zoom). The Teacher 1 appears as a host with the name correct in the Zoom app (not need to login). 

Same case with Teacher 2 (PRO ZOOM)

It doesn't work if the Teacher 1 or Teacher 2 are co-host. And there ir not logical reason for that unless Moodle is not sending the right co-host information to Zoom. 

Also it works using the LTI connection (same premises, same users).

:)

**Comment by jrchamp** [comment]
Hi @SnacksonMA 

If it works one way, but not the other, then I need more information. The two pieces that I can think of are:

1. Do the meeting details look the same on the Zoom portal? If not, what are the differences?
2. Are the join URLs the same? If not, what are the differences?

If the meeting configuration is the same on the Zoom portal and the URL to join the meeting is the same, then it should all behave the same.

I looked at your comment where you said:

> https://alku-videoclases-es.zoom.us/j/NNNNNNNNNNN?uname=Profe%20Zigurat#success

I checked the history of the code and the "uname" query parameter was added in the initial commit, roughly 6 years ago. Nowhere in the current, official Zoom documentation is that query parameter mentioned. Some searching indicates that it was deprecated and removed a while ago. It's possible that this parameter is causing a problem and should be removed.

If you would be willing to try copying the URL, removing the `uname=Profe%20Zigurat` piece and joining that way, it would be helpful to know if that fixes the problem.
